File: Medium/62. Unique Paths/Solution.py
class Solution(object):
    def __init__(self):
        self.dp = []
    
    def out(self, m, n, i, j):
        return i < 0 or j < 0 or i >= m or j >= n; 

    def solve(self, m, n, i, j):
        if i == m - 1 and j == n - 1:
            return 1

        ret = self.dp[i][j]

        if ret != -1:
            return ret

        ret = 0
        if not self.out(m, n, i + 1, j):
            ret += self.solve(m, n, i + 1, j)
        if not self.out(m, n, i, j + 1):
            ret += self.solve(m, n, i, j + 1)

        self.dp[i][j] = ret
        return ret
    
    
    def uniquePaths(self, m, n):
        self.dp = [[-1 for col in range(n)] for row in range(m)]
        return self.solve(m, n, 0, 0)

File: Medium/62. Unique Paths/Solution.cpp
class Solution {
private:
    int dp[105][105];
public:
    Solution() { memset(dp, -1, sizeof dp); }
    bool out(int m, int n, int i, int j) { return i < 0 or j < 0 or i >= m or j >= n; }
    int solve(int m, int n, int i = 0, int j = 0) {
        if(i == m - 1 and j == n - 1) return 1;

        auto & ret = dp[i][j];
        if(~ret) return ret;

        ret = 0;
        if(!out(m, n, i + 1, j)) ret += solve(m, n, i + 1, j);
        if(!out(m, n, i, j + 1)) ret += solve(m, n, i, j + 1);
        return ret;
    }
    int uniquePaths(int m, int n) {
        return solve(m, n);
    }
};